Datasets modularization pt.5

nikpodsh opened the pull request:
### Feature or Bugfix
- Refactoring

### Detail
Refactoring of the Dataset entity and related to it code.
Refactoring for Votes
Introduced DataPolicy (the same way as ServicePolicy was used used)
Extracted dataset related permissions. 
Used new `has_tenant_permission` instead of `has_tenant_perm` that allows not to pass unused parameters

### Relates
#412 and #295

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

dbalintx commented on backend/dataall/api/Objects/Vote/resolvers.py:
```diff
@@ -49,3 +51,6 @@ def get_vote(context: Context, source, targetUri: str = None, targetType: str =
         )
+
+
+add_vote_type("dashboard", DashboardIndexer)
```
I'd leave here a TODO note to move this to the Dashboard module as soon as it gets created

nikpodsh replied:
Sure! Why not :)

dbalintx commented on backend/dataall/cdkproxy/stacks/policies/data_policy.py:
```diff
@@ -102,7 +103,7 @@ def generate_data_access_policy(self) -> iam.Policy:
         return policy
 
-    def get_statements(self):
+    def get_statements(self, session):
```
session is only used in the child class method, maybe replace it on this abstract level with args/kwargs?

dbalintx commented on backend/dataall/modules/datasets/services/share_notification_service.py:
```diff
@@ -0,0 +1,101 @@
+from dataall.modules.datasets.db.models import Dataset
+
+
+class ShareNotificationService:
```
As this class only has static methods, shouldn't these functions be on module level? (without a class)

I'd also rename the file to follow the pattern in the folder: dataset_ ...

nikpodsh replied:
I don't think that we will gain more benefits by removing the class. Yeah, it doesn't carry any functional role here, but it groups the code with the same functionality.

About the renaming: yes, I completely agree with you. This class will be changed when the sharing module comes up :)

dbalintx commented on backend/dataall/modules/datasets/services/permissions.py:
I'd rename the file to dataset_permissions.py

dbalintx commented on backend/dataall/modules/datasets/tasks/subscription_service.py:
Would it make sense to rename the file and the class to avoid confusion? It is scheduled as a task, but named as a service for which we have now a separate folder.

nikpodsh replied:
Yes, let's rename it!

dbalintx commented on backend/dataall/modules/datasets/services/dataset_group_resource.py:
```diff
@@ -0,0 +1,8 @@
+from dataall.modules.datasets.db.dataset_repository import DatasetRepository
+
+
+class DatasetGroupResource(GroupResource):
```
This could be moved to dataset_repository

dbalintx commented on backend/dataall/core/group/services/group_resource_manager.py:
```diff
@@ -0,0 +1,24 @@
+from typing import Protocol, List
+
+
+class GroupResource(Protocol):
```
As we agreed: this could be abstract class instead of a Protocol

dbalintx commented on backend/dataall/modules/datasets/services/dataset_service.py:
```diff
@@ -623,3 +619,82 @@ def count_dataset_tables(session, dataset_uri):
         )
+
+    @staticmethod
+    def query_environment_group_datasets(session, envUri, groupUri, filter) -> Query:
```
These query functions should be in dataset_repository (following your design pattern), but as we discussed this could also be in a followup :)

nikpodsh replied:
Yes, I will extract this methods after migrating the sharing.

dlpzx commented on backend/dataall/api/Objects/Stack/stack_helper.py:
```diff
@@ -84,15 +83,6 @@ def deploy_stack(targetUri):
         return stack
 
 
-def deploy_dataset_stack(dataset: models.Dataset):
```
I agree on moving this into the resolvers, it did not make much sense here

dlpzx commented on backend/dataall/modules/datasets/tasks/bucket_policy_updater.py:
Unrelated to modularization: this task needs to be reviewed. It is triggered every 15 mins and it was originally conceived to add statements to the pre-existing bucket policy of imported datasets. Through this task we can enforce more restrictive access to imported buckets

nikpodsh replied:
Yes, it looks strange. The task runs every 15 mins to update  policies for imported datasets. Most of the time it will do nothing, but just incur cost for running. 
I wonder why we can't update the policies during the importing of datasets?

dlpzx commented on backend/dataall/modules/datasets/aws/lf_table_client.py:
```diff
@@ -22,21 +24,36 @@ def grant_pivot_role_all_table_permissions(self):
         :param table:
```
:param table: not needed

dlpzx commented on backend/dataall/modules/datasets/aws/lf_table_client.py:
```diff
@@ -22,21 +24,36 @@ def grant_pivot_role_all_table_permissions(self):
         :param table:
```
Looking at the `clients` I see your logic of dividing the clients by service (glue, lf) and by dataset-object (dataset, table, storage_location...). At first I thought that it would be easier to have a single service client, but I think this approach by service+ by object makes it easier to follow and allows us to scale more easily. The only duplicated lines would be the "creating the client" lines from 13 to 16 if there are other lf_something_clients, but we save on parameters as having the _table as part of the class, so I would say it pays off. 

The only thing to add, there is some sort of `lf_dataset_client` in the creation of datasets. In `backend/dataall/cdkproxy/assets/gluedatabasecustomresource` This is a custom resource Lambda that uses some LakeFormation methods.

dlpzx commented on backend/dataall/modules/datasets/handlers/glue_column_handler.py:
```diff
@@ -38,7 +38,7 @@ def update_table_columns(engine, task: models.Task):
 
             aws_session = SessionHelper.remote_session(table.AWSAccountId)
 
```
minor suggestion: define params`table=table, aws_session=aws_session` to avoid mistakes

dlpzx commented on backend/dataall/modules/datasets/handlers/glue_dataset_handler.py:
If it is going to contain only the crawler maybe we can rename the file to something that indicates that it is containing the Crawling handling

dlpzx commented on backend/dataall/modules/datasets/handlers/glue_table_handler.py:
```diff
@@ -18,7 +17,7 @@ class DatasetColumnGlueHandler:
     @Worker.handler(path='glue.dataset.database.tables')
     def list_tables(engine, task: models.Task):
```
This method has a very bad name, not a problem of modularization but the function is actually syncing tables from glue catalog to data.all. And the handler should be called something like glue_table_sync_handler

dlpzx commented on backend/dataall/modules/datasets/handlers/sns_dataset_handler.py:
In line 30 and 40 we are creating and using the sns client, would this not be part of clients? Something like 'sns_dataset_client' and then in the handlers we would have something like dataset_sns_subscription_handler

nikpodsh replied:
Yep, I had thought to migrate it later, but since you have pointed it out I will do it here

dlpzx commented on backend/dataall/modules/datasets/handlers/s3_location_handler.py:
Same as other handlers, if we are creating separate files I would go for more precise names. This is related to creation of folders

dlpzx commented on backend/dataall/api/Objects/Environment/resolvers.py:
```diff
@@ -370,23 +370,6 @@ def list_environment_group_permissions(
-        )
-
-
 def list_shared_with_environment_data_items(
```
I guess that `list_shared_with_environment_data_items` will also be deleted from here and replaced by the `count` in GroupResources once shares are modularized right?

dlpzx commented on backend/dataall/aws/handlers/glue.py:
```diff
@@ -370,142 +368,6 @@ def batch_delete_tables(**data):
-            )
-
-    @staticmethod
-    def create_glue_crawler(**data):
```
What happened with this method? I do not see it in glue_dataset_client, was it unused?

nikpodsh replied:
This was the dead code. It should have been been invoked by sending SQS action `glue.dataset.crawler.create`, but there was no `glue.dataset.crawler.create` anymore in the project, I deleted it

dlpzx commented on backend/dataall/cdkproxy/stacks/policies/data_policy.py:
```diff
@@ -86,11 +84,14 @@ def generate_admins_data_access_policy(self) -> iam.Policy:
 
-    def generate_data_access_policy(self) -> iam.Policy:
+    def generate_data_access_policy(self, session) -> iam.Policy:
         """
```
why do we need to pass the session?

dlpzx commented on backend/dataall/db/api/share_object.py:
```diff
@@ -561,7 +563,9 @@ def submit_share_object(
 
         Share_SM.update_state(session, share, new_share_state)
 
```
ok!

dlpzx commented on backend/dataall/modules/datasets/cdk/dataset_data_policy.py:
```diff
@@ -0,0 +1,33 @@
+from dataall.modules.datasets.services.dataset_service import DatasetService
+
+
+class DatasetDataPolicy(DataPolicy):
```
Maybe we can rename it S3 policy as it is what is actually doing and then everything is a service policy

dlpzx replied:
Right now "data" can be anything

dlpzx commented on tests/api/test_dataset_location.py:
```diff
@@ -70,10 +73,8 @@ def test_get_dataset(client, dataset1, env1, user, group):
 
 
```
What is the advantage of using MagicMock?

nikpodsh replied:
Honestly? it's just working :) 
I couldn't find a better way to mock _instance_ of class. Tried a few other ways to patch an instance, but got only `FAILED`. This one worked)

dlpzx reviewed (commented):
thank you!

nikpodsh commented:
Added a new commits:
Most of them were review remarks, but I also returned triggering the alarms as we discussed (+ added a line to trigger alarm not only create a message) @dlpzx 
Extracted DatasetRole (kudos to @dbalintx for noticing it)
Created a way to extend `EnvironmentSetup` (which is an environment stack) and extracted glue profiler since it looks related to datasets. I didn't do that for the others like`lakeformationdefaultsettings` and `gluedatabasecustomresource` since wasn't sure if there are only dataset-specific.
